include/pstore/core/hamt_map.hpp

```diff
@@ -457,6 +458,27 @@ namespace pstore {
             /// internal or linear node.
             void delete_node (index_pointer node, unsigned shifts);
 
+            static constexpr auto internal_nodes_per_chunk =
+                std::size_t{256} * 1024 / sizeof (internal_node);
+
+            /// Internal nodes are allocated using a "chunked-vector". This allocates memory in
+            /// lumps sufficent for internal_nodes_per_chunk entries. This is then consumed as new
+            /// in-heap internal nodes are created.
+            ///
+            /// The effect of this is to significantly reduce the number of memory allocations
+            /// performed by the index code under heavy insertion pressure as thus give us a little
+            /// extra performance. The cost is that we may waste as much as one chunk's worth of
+            /// memory (minus sizeof (internal_node).
+            ///
+            /// TODO: we allocate nodes at their maximum size even though they may only contain two
+            /// members. This is rather wasteful and will prevent us from moving to larger hash
+            /// sizes due to the bloated memory consumption.
+            using internal_nodes_container =
+                chunked_vector<internal_node, internal_nodes_per_chunk,
+                               sizeof (internal_node) +
+                                   sizeof (index_pointer) * sizeof (hash_type) * 8U - 1U>;
```

Comment:
Why don't use the `internal_node::size_bytes` function to give the `ActualSize`?

Reply:
> Why don't use the `internal_node::size_bytes` function to give the `ActualSize`?

Good point. I’ll have to make it constexpr to use it in a compile-time constant, but nonetheless…

Having said that, the assertion in that function is curious:
~~~c++
static std::size_t size_bytes (std::size_t const num_children) noexcept {
    assert (num_children > 0 && num_children < max_hash_bits);
    return sizeof (internal_node) - …
}
~~~
Isn’t `max_hash_bits` wrong? Shouldn’t it be `hash_size`? (The maximum number of children that an internal_node can contain is 64 ATM rather than 66.)

Reply:
> In the file include/pstore/adt/chunked_vector.hpp:
> Lines 87 and 88 should be deleted if the function will not be implemented in the future.

Who can say what the future will bring! They are there because the container should have reverse_iterator. I don’t need it right now so cannot justify the cost of implementing it right now.

> The function implementation of reference front () (line 127) and const_reference front () const (line 133) are quite duplicated. Could it be avoided? The same for the reference back () and const_reference back () const

The duplication in each case is one line of code plus an assertion. This can be avoided but at the cost of a significant amount of boilerplate. Sharing implementations of these const and non-const member functions isn’t trivial: see `inherit_const` and the large comment in inherit_const.hpp for the gory details. IMO, the extra noise from the boilerplate is not worth it for so little code.

> Before Line 367, should add an assert (It > 0)?

There’s no lt here: do you mean `it_->size () >0`? That certainly wouldn’t hurt.

Reply:
>
> Isn’t max_hash_bits wrong? Shouldn’t it be hash_size? (The maximum number of children that an internal_node can contain is 64 ATM rather than 66.)
>
You are right. Could you please fix the assert as well.